<div style="border:solid green 2px; padding: 20px">

Привет, Павел 🖐

Меня зовут Саша Данилевский, я буду делать ревью твоей работы по базовому Python. Если не против, давай будем общаться на «ты».
<br><br>

Ты уже проделал большую работу над проектом, но давай сделаем его еще лучше.
Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать хорошим специалистом.
    
Давай зададим цветовые метки для комментариев:

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Так я буду выделять критические ошибки, требующие обязательного исправления. Без них проект не может быть принят; </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Так я буду выделять то, что в следующий раз можно сделать по-другому. Один-два таких небольших замечаний в проекте допустимы, но если их будет больше — тебе будет необходимо внести исправления;</div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Так я буду отмечать крутые решения, на которые можно опираться в будущих проектах.</div>
    
Хочу попросить тебя: если **что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Используй для своих комментариев какой-либо цвет, отличный от используемых мной, так мне будет легче отследить изменения. Например, так:
    
<div class="alert alert-info"> <b>Комментарий студента:</b> текст комментария</div>
    
Пожалуйста, **не перемещай**, **не изменяй** и **не удаляй** мои комментарии, чтобы повторная проверка твоего проекта прошла оперативнее.
    
🙏 Извини за долгое ожидание ревью. При следующих проверках SLA будет соблюдаться.
</div>

<div class="alert alert-info"> <b>Комментарий студента:</b>
    
Привет, Саша! Рад знакомству! <br /> Спасибо за вдохновляющее вступпительное слово. В школе и в университете нам часто указывали на ошибки и часто за них ругаали. По старой привычке присуствует воленние, когда открываешь тетрадку с ревью. Но я приятно удивлён твоим подходом и подходом Яндекс.Практикума :)</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2✔️:</b>
    
Но ты ведь действительно круто выполнил проект
</div>

# Яндекс.Музыка

Сравнение Москвы и Петербурга окружено мифами. Например:
 * Москва — мегаполис, подчинённый жёсткому ритму рабочей недели;
 * Петербург — культурная столица, со своими вкусами.

На данных Яндекс.Музыки вы сравните поведение пользователей двух столиц.

**Цель исследования** — проверьте три гипотезы:
1. Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному.
2. В понедельник утром в Москве преобладают одни жанры, а в Петербурге — другие. Так же и вечером пятницы преобладают разные жанры — в зависимости от города. 
3. Москва и Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Петербурге — русский рэп.

**Ход исследования**

Данные о поведении пользователей вы получите из файла `yandex_music_project.csv`. О качестве данных ничего не известно. Поэтому перед проверкой гипотез понадобится обзор данных. 

Вы проверите данные на ошибки и оцените их влияние на исследование. Затем, на этапе предобработки вы поищете возможность исправить самые критичные ошибки данных.
 
Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Проверка гипотез.



**Учебная цель исследования**: повторить и закрепить пройденный материал по работе Pandas.

**Практическая цель исследования:** проверить три вышеуказанные гипотезы. 
Полученный пайплайн, с внесением доработок, может быть поелзен 
при обработке данных для других городов и полной недели.



## Обзор данных

Составьте первое представление о данных Яндекс.Музыки.




Основной инструмент аналитика — `pandas`. Импортируйте эту библиотеку.

In [1]:
# импорт библиотеки pandas
import pandas as pd

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Отлично! Очень хорошо, что ты используешь сокращение `pd` для библиотеки `pandas`. Это очень удобно для дальнейшей работы. Это относится и к другим библиотекам, которые ты будешь использовать в будущем.
</div>

Прочитайте файл `yandex_music_project.csv` из папки `/datasets` и сохраните его в переменной `df`:

In [2]:
# чтение файла с данными и сохранение в df
df = pd.read_csv ('/datasets/yandex_music_project.csv')

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Верно, слэш в начале пути важен, так как указывает, что нужно искать файл в корневой папке. </div>

Выведите на экран первые десять строк таблицы:

In [3]:
# получение первых 10 строк таблицы df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                    Преданная         IMPERVTOR  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE     И вновь продолжается бой               NaN  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

             City        time        Day  
0  Saint-Petersburg  20:28:33  Wednesday  
1            Moscow  14:07:09     Friday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
4            Moscow  08:34:34     Monday  
5  Saint-Petersburg  13:09:41     Friday  
6            Moscow  13:00:07  Wednesday  
7            Moscow  20:47:49  Wednesday  
8            Moscow  09:17:40     Friday  
9  Saint-Petersburg  21:20:49  Wednesday

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Верное использование `head()`.
А еще есть аналогичный метод `tail()`, который возвращает последние строки датафрейма.

`display()` здесь можно не использовать: последняя (а в нашем случае единственная) строка ячейки будет выведена сама.</div>

Одной командой получить общую информацию о таблице:

In [4]:
df.info() # получим общую информацию о таблице df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
  userID    65079 non-null object
Track       63848 non-null object
artist      57876 non-null object
genre       63881 non-null object
  City      65079 non-null object
time        65079 non-null object
Day         65079 non-null object
dtypes: object(7)
memory usage: 3.5+ MB


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Именно, `info()` позволяет самым удобным и простым способом посмотреть подробную информацию о датафрейме.</div>

Итак, в таблице семь столбцов. Тип данных во всех столбцах — `object`.

Согласно документации к данным:
* `userID` — идентификатор пользователя;
* `Track` — название трека;  
* `artist` — имя исполнителя;
* `genre` — название жанра;
* `City` — город пользователя;
* `time` — время начала прослушивания;
* `Day` — день недели.

В названиях колонок видны три нарушения стиля:
1. Строчные буквы сочетаются с прописными.
2. Встречаются пробелы.
3. Найдите ещё одну проблему в названии колонок и опишите её в этом пункте.

Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.


**Мой комментарий**

Имеется слитная запись и "верблюжий стиль" - в столбце `userID` 

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>

Совершенно верно, не используется «змеиный» регистр.

Ответ можно дать прям вместо формулировки.</div>

**Выводы**

В каждой строке таблицы — данные о прослушанном треке. Часть колонок описывает саму композицию: название, исполнителя и жанр. Остальные данные рассказывают о пользователе: из какого он города, когда он слушал музыку. 

Предварительно можно утверждать, что, данных достаточно для проверки гипотез. Но встречаются пропуски в данных, а в названиях колонок — расхождения с хорошим стилем. 


Чтобы двигаться дальше, нужно устранить проблемы в данных.

## Предобработка данных
Исправьте стиль в заголовках столбцов, исключите пропуски. Затем проверьте данные на дубликаты.

Проверку будем проводять для явных и скрытых дубликатов.

### Стиль заголовков
Выведите на экран названия столбцов:

In [5]:
# перечень названий столбцов таблицы df
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Ты верно используешь `columns` для решения задачи. Отлично, что ты здесь не использовал функцию `print()`: результат и так будет выведен, ведь это последняя строчка ячейки.</div>

Приведите названия в соответствие с хорошим стилем:
* несколько слов в названии запишите в «змеином_регистре»,
* все символы сделайте строчными,
* устраните пробелы.

Для этого переименуйте колонки так:
* `'  userID'` → `'user_id'`;
* `'Track'` → `'track'`;
* `'  City  '` → `'city'`;
* `'Day'` → `'day'`.

In [6]:
# переименование столбцов
df = df.rename (columns = {'  userID':'user_id', 'Track':'track', '  City  ':'city', 'Day':'day'})

Проверьте результат. Для этого ещё раз выведите на экран названия столбцов:

In [7]:
# проверка результатов после переименования столбцов
# перечень названий столбцов
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Хороший выбор метода, так как `rename()` сработает, даже если изменится порядок столбцов. А это будет удобно, когда столбцов для переименования станет намного больше.
</div>

### Пропуски значений
Сначала посчитайте, сколько в таблице пропущенных значений. Для этого достаточно двух методов `pandas`:

In [8]:
# подсчёт пропусков
df.isna().sum() # оценим количество пропусков 

user_id       0
track      1231
artist     7203
genre      1198
city          0
time          0
day           0
dtype: int64

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Хорошо, что для нахождения пропущенных значений ты выбрал метод `isna()`. На самом деле, у него есть альтернативы, которые делают одно и то же (например, `isnull()`), но использование `isna()` все же предпочтительнее.
</div>

Не все пропущенные значения влияют на исследование. Так в `track` и `artist` пропуски не важны для вашей работы. Достаточно заменить их явными обозначениями.

Но пропуски в `genre` могут помешать сравнению музыкальных вкусов в Москве и Санкт-Петербурге. На практике было бы правильно установить причину пропусков и восстановить данные. Такой возможности нет в учебном проекте. Придётся:
* заполнить и эти пропуски явными обозначениями,
* оценить, насколько они повредят расчётам. 

Замените пропущенные значения в столбцах `track`, `artist` и `genre` на строку `'unknown'`. Для этого создайте список `columns_to_replace`, переберите его элементы циклом `for` и для каждого столбца выполните замену пропущенных значений:

In [9]:
# перебор названий столбцов в цикле и замена пропущенных значений на 'unknown'
columns_to_replace = ['track', 'artist', 'genre']
for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')

Убедитесь, что в таблице не осталось пропусков. Для этого ещё раз посчитайте пропущенные значения.

In [10]:
# подсчёт пропусков
# оценим количество пропусков после заполнения пропущенных значений
df.isna().sum() 

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

Повторный подсчёт пропусков показывает, что все пропущенные значения были заполнены

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Решение составлено совершенно верно, цикл отрабатывает без ошибок.
</div>

### Дубликаты
Посчитайте явные дубликаты в таблице одной командой:

In [11]:
# подсчёт явных дубликатов
ammount_dulicates = df.duplicated().sum()
print ('Количесвто явных дубликатов: ', ammount_dulicates)

Количесвто явных дубликатов:  3826


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Дубликаты найдены и подсчитаны верным и оптимальным способом.
</div>

Вызовите специальный метод `pandas`, чтобы удалить явные дубликаты:

In [12]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index (drop = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61253 entries, 0 to 61252
Data columns (total 7 columns):
user_id    61253 non-null object
track      61253 non-null object
artist     61253 non-null object
genre      61253 non-null object
city       61253 non-null object
time       61253 non-null object
day        61253 non-null object
dtypes: object(7)
memory usage: 3.3+ MB


До удаления дубликатов - 65079 строк.

После удаления дубликатов - 61253 строки. 



Ещё раз посчитайте явные дубликаты в таблице — убедитесь, что полностью от них избавились:

In [13]:
# повторная проверка на отсутствие дубликатов
ammount_dulicates = df.duplicated().sum()
print ('Количесвто явных дубликатов после удаления: ', ammount_dulicates)

Количесвто явных дубликатов после удаления:  0


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Верно! Хорошо, что умеешь удалять старые индексы с помощью параметра `drop`.
</div>

Теперь избавьтесь от неявных дубликатов в колонке `genre`. Например, название одного и того же жанра может быть записано немного по-разному. Такие ошибки тоже повлияют на результат исследования.

Выведите на экран список уникальных названий жанров, отсортированный в алфавитном порядке. Для этого:
* извлеките нужный столбец датафрейма, 
* примените к нему метод сортировки,
* для отсортированного столбца вызовите метод, который вернёт уникальные значения из столбца.

In [14]:
# Просмотр уникальных названий жанров

# уникальные значения столбца "жанр", отсортированные в алфавитном порядке
sorted_gerne_series = df['genre'].sort_values().unique() 
print ('Количество уникальных жанров до удаления дубликатов: ', len (sorted_gerne_series))
sorted_gerne_series

Количество уникальных жанров до удаления дубликатов:  290


array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Список совсем не маленький — наверняка многие названия обозначают один и тот же жанр.
</div>

Просмотрите список и найдите неявные дубликаты названия `hiphop`. Это могут быть названия с ошибками или альтернативные названия того же жанра.

Вы увидите следующие неявные дубликаты:
* *hip*,
* *hop*,
* *hip-hop*.

Чтобы очистить от них таблицу, напишите функцию `replace_wrong_genres()` с двумя параметрами: 
* `wrong_genres` — список дубликатов,
* `correct_genre` — строка с правильным значением.

Функция должна исправить колонку `genre` в таблице `df`: заменить каждое значение из списка `wrong_genres` на значение из `correct_genre`.

In [15]:
# Функция для замены неявных дубликатов
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace (wrong_genre, correct_genre)

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Функция написана верно.
</div>

Вызовите `replace_wrong_genres()` и передайте ей такие аргументы, чтобы она устранила неявные дубликаты: вместо `hip`, `hop` и `hip-hop` в таблице должно быть значение `hiphop`:

In [16]:
# Устранение неявных дубликатов 'hiphop'
wrong_genres = ['hip', 'hip-hop', 'hop']
correct_genre = 'hiphop'
                
replace_wrong_genres(wrong_genres, correct_genre)

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> 
    
Отлично!
</div>

Проверьте, что заменили неправильные названия:

*   hip
*   hop
*   hip-hop

Выведите отсортированный список уникальных значений столбца `genre`:

In [17]:
# Проверка на неявные дубликаты после удаления hiphop
sorted_gerne_series_clean = df['genre'].sort_values().unique() 
print ('Количество уникальных жанров до удаления дубликатов: ', len (sorted_gerne_series))
print ('Количество уникальных жанров после удаления дубликатов: ', len (sorted_gerne_series_clean))
sorted_gerne_series_clean

Количество уникальных жанров до удаления дубликатов:  290
Количество уникальных жанров после удаления дубликатов:  287


array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Отлично, дублей хип-хопа больше нет.
</div>

**При более внимательном изучении данных были ещё найдены неявные дубликаты:**

- latin ('latin', 'latino');
- turkey ('türk', 'türkçe');
- электроника ('electronic').

Удалим вышеописанные дубликаты

In [18]:
# Устранение неявных дубликатов - latin ('latin', 'latino');
wrong_genres = ['latin']
correct_genre = 'latino'
replace_wrong_genres(wrong_genres, correct_genre)

# Устранение неявных дубликатов - turkey ('türk', 'türkçe');
wrong_genres = ['türk', 'türkçe']
correct_genre = 'turkey'
replace_wrong_genres(wrong_genres, correct_genre)

# Устранение неявных дубликатов - электроника ('electronic').
wrong_genres = ['электроника']
correct_genre = 'electronic'              
replace_wrong_genres(wrong_genres, correct_genre)

In [19]:
# Проверка на неявные дубликаты после тщательного удаления (latin, turkey, электроника)

sorted_gerne_series_clean = df['genre'].sort_values().unique() 
print ('Количество уникальных жанров до удаления дубликатов: ', len (sorted_gerne_series))
print ('Количество уникальных жанров после тщательного удаления дубликатов: ', len (sorted_gerne_series_clean))
sorted_gerne_series_clean

Количество уникальных жанров до удаления дубликатов:  290
Количество уникальных жанров после тщательного удаления дубликатов:  284


array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

**Выводы**

Предобработка обнаружила три проблемы в данных:

- нарушения в стиле заголовков,
- пропущенные значения,
- дубликаты — явные
- **неявные (hiphop, latino, turkey, electronic)**

Вы исправили заголовки, чтобы упростить работу с таблицей. Без дубликатов исследование станет более точным.

Пропущенные значения вы заменили на `'unknown'`. Ещё предстоит увидеть, не повредят ли исследованию пропуски в колонке `genre`.

Теперь можно перейти к проверке гипотез. 

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Верно! Ты молодец, что выходишь за рамки задания
</div>

## Проверка гипотез

### Сравнение поведения пользователей двух столиц

Первая гипотеза утверждает, что пользователи по-разному слушают музыку в Москве и Санкт-Петербурге. Проверьте это предположение по данным о трёх днях недели — понедельнике, среде и пятнице. Для этого:

* Разделите пользователей Москвы и Санкт-Петербурга
* Сравните, сколько треков послушала каждая группа пользователей в понедельник, среду и пятницу.


Для тренировки сначала выполните каждый из расчётов по отдельности. 

Оцените активность пользователей в каждом городе. Сгруппируйте данные по городу и посчитайте прослушивания в каждой группе.



In [20]:
# Подсчёт прослушиваний в каждом городе
group_by_city = df.groupby('city')['track'].count()
group_by_city 

city
Moscow              42741
Saint-Petersburg    18512
Name: track, dtype: int64

В Москве прослушиваний больше, чем в Петербурге. Из этого не следует, что московские пользователи чаще слушают музыку. Просто самих пользователей в Москве больше.

Теперь сгруппируйте данные по дню недели и подсчитайте прослушивания в понедельник, среду и пятницу. Учтите, что в данных есть информация только о прослушиваниях только за эти дни.


In [21]:
# Подсчёт прослушиваний в каждый из трёх дней
group_by_day = df.groupby('day')['track'].count()
group_by_day

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Отлично. Здесь можно считать количество по любому столбцу, так количество значений везде одинаковое.
</div>

В среднем пользователи из двух городов менее активны по средам. Но картина может измениться, если рассмотреть каждый город в отдельности.

Вы видели, как работает группировка по городу и по дням недели. Теперь напишите функцию, которая объединит два эти расчёта.

Создайте функцию `number_tracks()`, которая посчитает прослушивания для заданного дня и города. Ей понадобятся два параметра:
* день недели,
* название города.

В функции сохраните в переменную строки исходной таблицы, у которых значение:
  * в колонке `day` равно параметру `day`,
  * в колонке `city` равно параметру `city`.

Для этого примените последовательную фильтрацию с логической индексацией.

Затем посчитайте значения в столбце `user_id` получившейся таблицы. Результат сохраните в новую переменную. Верните эту переменную из функции.

In [22]:
# Функция для подсчёта прослушиваний для конкретного города и дня
# С помощью последовательной фильтрации с логической индексацией она 
# сначала получит из исходной таблицы строки с нужным днём и городом (track_list)
# методом count() посчитает количество значений в колонке user_id. (track_list_count) 
# Это количество функция вернёт в качестве результата (return track_list_count)

# <создание функции number_tracks()>
def number_tracks(day, city):  # Объявляется функция с двумя параметрами: day, city.
    track_list = df[(df['day'] == day) & (df['city'] == city)]
    track_list_count = track_list['user_id'].count()
    return track_list_count

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Отличная функция.
    
Только советую следить за единообразием стиля твоего кода: где-то у тебя есть отступы, где-то нет.
</div>

<div class="alert alert-warning">
<s><b>Комментарий ревьюера ⚠️:</b>
    
Комментарии лучше оставлять НАД кодом. Так твой проект будет удобней и проще читать, будет сразу понятно, что делает твоя функция. Но это относится не только к коду функций.
</s></div>

<div class="alert alert-info"> <b>Комментарий студента:</b>
    
Спасибо за замечания по поводу однообразия отступов и про помещение кода перед функцией. Я поправил :)</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2✔️:</b>
    
👍
</div>

Вызовите `number_tracks()` шесть раз, меняя значение параметров — так, чтобы получить данные для каждого города в каждый из трёх дней.

**Автоматизируем процесс и создадим цикл для этой задачи**

In [23]:
cities = ['Moscow', 'Saint-Petersburg'] # список городов
days = ['Monday', 'Wednesday', 'Friday'] # список дней недели

data_set_for_days = [] # создадим пустой список для сохраненния результатов в нём

for day in days:
    list_for_city = [] # создадим пустой список для записи результатов для каждого дня недели
    for city in cities:    
        list_for_city.append (number_tracks(day, city))
    data_set_for_days.append (list_for_city) 

data_set_for_days

[[15740, 5614], [11056, 7003], [15945, 5895]]

Создайте c помощью конструктора `pd.DataFrame` таблицу, где
* названия колонок — `['city', 'monday', 'wednesday', 'friday']`;
* данные — результаты, которые вы получили с помощью `number_tracks`.

In [24]:
# Таблица с результатами
data_city_day = pd.DataFrame (data = data_set_for_days, columns = cities, index=days)
data_city_day

Moscow  Saint-Petersburg
Monday      15740              5614
Wednesday   11056              7003
Friday      15945              5895

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Все верно посчитано и перенесено в датафрейм.  Очень крутая автоматизация! Это точно поможет избежать возможных ошибок в будущем: например, если изменятся исходные данные, не придется все вручную исправлять.
    
Хотя в задании строки и столбцы должны быть поменяны местами, твое решение я точно принимаю 👍
    
Лайфхак: если нужно перевернуть датафрейм (транспонировать), можно сделать так: `data_city_day.T`
</div>

<div class="alert alert-info"> <b>Комментарий студента:</b>
    
Спасибо за лайфак про транспонирование. Ниже я привел таблику в соответствие </div>

In [25]:
data_city_day.T 

Monday  Wednesday  Friday
Moscow             15740      11056   15945
Saint-Petersburg    5614       7003    5895

<div class="alert alert-success">
<b>Комментарий ревьюера v2✔️:</b>
    
Да, так круто.
</div>

**Выводы**

Данные показывают разницу поведения пользователей:

- В Москве пик прослушиваний приходится на понедельник и пятницу, а в среду заметен спад.
- В Петербурге, наоборот, больше слушают музыку по средам. Активность в понедельник и пятницу здесь почти в равной мере уступает среде.

Значит, данные говорят в пользу первой гипотезы. 

Активность пользователей зависит от дня недели. Причём для Москвы и Санкт-Петербурга это проявляется по-разному

### Музыка в начале и в конце недели

Согласно второй гипотезе, утром в понедельник в Москве преобладают одни жанры, а в Петербурге — другие. Так же и вечером пятницы преобладают разные жанры — в зависимости от города.

Сохраните таблицы с данными в две переменные:
* по Москве — в `moscow_general`;
* по Санкт-Петербургу — в `spb_general`.

In [26]:
# получение таблицы moscow_general из тех строк таблицы df, для которых значение в столбце 'city' равно 'Moscow'
# получение таблицы spb_general из тех строк таблицы df, для которых значение в столбце 'city' равно 'Saint-Petersburg'

moscow_general = df[df['city'] == 'Moscow']
spb_general = df[df['city'] == 'Saint-Petersburg']

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Ok.
</div>

Создайте функцию `genre_weekday()` с четырьмя параметрами:
* таблица (датафрейм) с данными,
* день недели,
* начальная временная метка в формате 'hh:mm', 
* последняя временная метка в формате 'hh:mm'.

Функция должна вернуть информацию о топ-10 жанров тех треков, которые прослушивали в указанный день, в промежутке между двумя отметками времени.

In [27]:
def genre_weekday(dataframe, day, time_1, time_2): # Объявление функции genre_weekday() с параметрами table, day, time_1, time_2
    genre_df = dataframe[dataframe['day'] == day]  # фильтрация по дню недели
    genre_df = genre_df [(genre_df['time'] > time_1) & (genre_df['time'] < time_2)]  # фильтрация по времени
    genre_df_count = genre_df.groupby('genre')['track'].count() # группировка по жанру и подсчёт количесвта треков
    genre_df_count_sorted =  genre_df_count.sort_values(ascending = False) # сортировка результата 
    return genre_df_count_sorted [:10]   

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Отлично, все требования к функции выполнены.
    
Только следи за тем, чтобы комментарии совпадали с кодом. Сейчас имена атрибутов расходятся.
</div>

<div class="alert alert-info"> <b>Комментарий студента:</b>
    
Спасибо за замечание про расхождение имён атрибутов в коде и в комментариях. Имена   *time_1, time_2*   поправил </div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2✔️:</b>
    
👍
</div>

In [28]:
# Мсоква. Понедельник. Утро
genre_weekday (moscow_general, 'Monday', '07:00', '11:00')

genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
Name: track, dtype: int64

Cравните результаты функции `genre_weekday()` для Москвы и Санкт-Петербурга в понедельник утром (с 7:00 до 11:00) и в пятницу вечером (с 17:00 до 23:00):

In [29]:
# Спб. Понедельник. Утро
genre_weekday (spb_general, 'Monday', '07:00', '11:00')

genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
Name: track, dtype: int64

In [30]:
# Мсоква. Пятница. Вечер
genre_weekday (moscow_general, 'Friday', '17:00', '23:00')

genre
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world          208
ruspop         170
classical      163
alternative    163
rusrap         142
Name: track, dtype: int64

In [31]:
# Спб. Пятница. Вечер
genre_weekday (spb_general, 'Friday', '17:00', '23:00')

genre
pop            256
rock           216
electronic     216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
Name: track, dtype: int64

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Да, с помощью функции мы получили всю нужную информацию для того, чтобы сделать выводы.
</div>

**Выводы**

Если сравнить топ-10 жанров в понедельник утром, можно сделать такие выводы:

1. В Москве и Петербурге слушают похожую музыку. Единственное отличие — в московский рейтинг вошёл жанр “world”, а в петербургский — джаз и классика.

2. В Москве пропущенных значений оказалось так много, что значение `'unknown'` заняло десятое место среди самых популярных жанров. Значит, пропущенные значения занимают существенную долю в данных и угрожают достоверности исследования.

Вечер пятницы не меняет эту картину. Некоторые жанры поднимаются немного выше, другие спускаются, но в целом топ-10 остаётся тем же самым.

Таким образом, вторая гипотеза подтвердилась лишь частично:
* Пользователи слушают похожую музыку в начале недели и в конце.
* Разница между Москвой и Петербургом не слишком выражена. В Москве чаще слушают русскую популярную музыку, в Петербурге — джаз.

Однако пропуски в данных ставят под сомнение этот результат. В Москве их так много, что рейтинг топ-10 мог бы выглядеть иначе, если бы не утерянные  данные о жанрах.

### Жанровые предпочтения в Москве и Петербурге

Гипотеза: Петербург — столица рэпа, музыку этого жанра там слушают чаще, чем в Москве.  А Москва — город контрастов, в котором, тем не менее, преобладает поп-музыка.

Сгруппируйте таблицу `moscow_general` по жанру и посчитайте прослушивания треков каждого жанра методом `count()`. Затем отсортируйте результат в порядке убывания и сохраните его в таблице `moscow_genres`.

In [32]:
# одной строкой: группировка таблицы moscow_general по столбцу 'genre', 
# подсчёт числа значений 'genre' в этой группировке методом count(), 
# сортировка получившегося Series в порядке убывания и сохранение в moscow_genres

moscow_genres = moscow_general.groupby('genre')['genre'].count().sort_values(ascending = False)


Выведите на экран первые десять строк `moscow_genres`:

In [33]:
# просмотр первых 10 строк moscow_genres
moscow_genres[:10]

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Данные по Москве получены верно.
</div>

Теперь повторите то же и для Петербурга.

Сгруппируйте таблицу `spb_general` по жанру. Посчитайте прослушивания треков каждого жанра. Результат отсортируйте в порядке убывания и сохраните в таблице `spb_genres`:


In [34]:
# одной строкой: группировка таблицы spb_general по столбцу 'genre', 
# подсчёт числа значений 'genre' в этой группировке методом count(), 
# сортировка получившегося Series в порядке убывания и сохранение в spb_genres

spb_genres = spb_general.groupby('genre')['genre'].count().sort_values(ascending = False)

Выведите на экран первые десять строк `spb_genres`:

In [35]:
# просмотр первых 10 строк spb_genres
spb_genres[:10]

genre
pop            2431
dance          1932
rock           1879
electronic     1737
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Данные по Петербургу тоже получены успешно.
</div>

**Выводы**

Гипотеза частично подтвердилась:
* Поп-музыка — самый популярный жанр в Москве, как и предполагала гипотеза. Более того, в топ-10 жанров встречается близкий жанр — русская популярная музыка.
* Вопреки ожиданиям, рэп одинаково популярен в Москве и Петербурге. 


## Особенности в данных

<div class="alert alert-info"> <b>Комментарий студента:</b>
    
В процессе исселдования возникла гипотеза:
Каждый пользователь слушал по одному треку каждого жанра. Никто не слушал, нарпимер, два трека Рэпа, или два трека Рока. <br /> Проверим гипотезу</div>

In [36]:
# сгруппируем данные по 'user_id' и выведем для каждого пользователя список уникальных треков
# и общее количество вхождений для треков (количество прослушиваний)
track_count = df.groupby('user_id').agg({'track': ['count', 'unique']})
track_count

track                                                   
         count                                             unique
user_id                                                          
10004220     3             [Aisha, Аиша – Гриша, The Age Of Self]
100173ED     1                 [Contiuation of the Last Judgment]
1001CE90     7  [Lennart, Hellraiser III: Hell On Earth - Hell...
1002410A     3  [Wagogo Soothing Song (Tanzania), Baba Nam Kev...
100261F9     1                                             [Jump]
...        ...                                                ...
FFF965B2     2                        [Coffee, Let Me Be The One]
FFF98D49     1                                         [Part XIV]
FFFBB52      4  [One More Night (As Made Famous By Maroon 5), ...
FFFD03A      1  [Five Nights at Freddy's: Night 3 (feat. NateW...
FFFF6AC5     1                                    [Tika's Groove]

[41748 rows x 2 columns]

<div class="alert alert-info"> <b>Комментарий студента:</b>
    
Теперь у нас есть список уникальных треков для каждого пользователя и количесвто прослушиваний. <br /> Подсчитаем общее количество прослушиваний и количество уникальных треков</div>


In [37]:
# длина списка уникальных треков для каждого пользователя
track_unique_ammount = track_count['track']['unique'].apply (len)

# суммарное количество уникальных треков для всех пользователей
track_unique_sum = track_unique_ammount.sum() 

# общее количество прослушиваний для всех пользователей 
track_all = track_count['track']['count'].sum()


# создадим табличку для отображения 
# в переменную data_df запишем полученные значения  
data_df = [track_unique_sum, track_all] 

# названия строк
index_df = ['Уникальные треки', 'Прослушивания']  

# создание таблички
unique_df = pd.DataFrame (data = data_df, index = index_df, columns = ['Количество'])
unique_df

Количество
Уникальные треки       59573
Прослушивания          61253

<div class="alert alert-info"> <b>Комментарий студента:</b>
    
Занчения гооврят о том, что общее количество прослушиваний больше, чем количество уникальных треков. Это значит, что некоторое треки были прослушаны более одного раза <br /> Гипотеза не подтвердилась</div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
    
Не очень понятно, что ты хотел сделать здесь. Три твои группировки дадут одинаковое значение, так как все колонки без пропусков и содержат одинаковое число элементов.
</div>

<div class="alert alert-info"> <b>Комментарий студента:</b>
    
И правда, я ошибся в первой редакции, подумав что действительно каждый слушал музыку по одному разу. Поддался когнитивному искажению и "подогнал резульат под желаемый". Спасибо за замечание.</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2✔️:</b>
    
Вот, теперь с комментариями понятно, что ты делал.
    
Посчитать количество уникальных треков и жанров для каждого пользователя можно было группировкой: `df[df['track']!='unknown'].groupby(['user_id', 'genre', 'track'], as_index=False).agg({'time': 'nunique'}).sort_values(by='time', ascending = False)`.
</div>

In [39]:
df.head()

user_id                        track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

               city      time        day  
0  Saint-Petersburg  20:28:33  Wednesday  
1            Moscow  14:07:09     Friday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
4            Moscow  08:34:34     Monday

In [81]:
"""display (df[df['track']!='unknown'].groupby(['user_id', 'genre', 'track']).agg({'time': 'nunique'}).sort_values(by='time', ascending = False))
display (df[df['track']!='unknown'].groupby(['user_id', 'genre', 'track']).agg({'track': 'count'}).loc['E2EA9731'])
display (df[df['track']!='unknown'].groupby(['user_id', 'genre', 'track']).agg({'time': 'nunique'}).sort_values(by='time', ascending = False).loc['B3AEE0E'])

display (df[df['track']!='unknown'].groupby(['user_id', 'track'], as_index=False).agg({'time': 'nunique'}).sort_values(by='time', ascending = False).loc[40279])
display (df[df['track']!='unknown'].groupby(['user_id', 'track'], as_index=False).agg({'time': 'nunique'}).sort_values(by='time', ascending = False))"""

"display (df[df['track']!='unknown'].groupby(['user_id', 'genre', 'track']).agg({'time': 'nunique'}).sort_values(by='time', ascending = False))\ndisplay (df[df['track']!='unknown'].groupby(['user_id', 'genre', 'track']).agg({'track': 'count'}).loc['E2EA9731'])\ndisplay (df[df['track']!='unknown'].groupby(['user_id', 'genre', 'track']).agg({'time': 'nunique'}).sort_values(by='time', ascending = False).loc['B3AEE0E'])\n\ndisplay (df[df['track']!='unknown'].groupby(['user_id', 'track'], as_index=False).agg({'time': 'nunique'}).sort_values(by='time', ascending = False).loc[40279])\ndisplay (df[df['track']!='unknown'].groupby(['user_id', 'track'], as_index=False).agg({'time': 'nunique'}).sort_values(by='time', ascending = False))"

In [52]:
df['user_id'].nunique()
display (df.nunique())
df[df['track']!='unknown'].nunique()

user_id    41748
track      47246
artist     43606
genre        284
city           2
time       20392
day            3
dtype: int64

user_id    41408
track      47245
artist     43606
genre        283
city           2
time       20308
day            3
dtype: int64

## Итоги исследования

Были проверены три гипотезы и установлено:

1. День недели по-разному влияет на активность пользователей в Москве и Петербурге. 

**Первая гипотеза полностью подтвердилась.**

2. Музыкальные предпочтения не сильно меняются в течение недели — будь то Москва или Петербург. Небольшие различия заметны в начале недели, по понедельникам:
* в Москве слушают музыку жанра “world”,
* в Петербурге — джаз и классику.

**Вторая гипотеза подтвердилась частично.**
Этот результат мог оказаться иным, если бы не пропуски в данных.

3. Во вкусах пользователей Москвы и Петербурга больше общего чем различий. Вопреки ожиданиям, предпочтения жанров в Петербурге напоминают московские.

**Третья гипотеза не подтвердилась**. 
Если различия в предпочтениях и существуют, на основной массе пользователей они незаметны.

**На практике исследования содержат проверки статистических гипотез.**
Из данных одного сервиса не всегда можно сделать вывод о всех жителях города.
Проверки статистических гипотез покажут, насколько они достоверны, исходя из имеющихся данных. 
С методами проверок гипотез вы ещё познакомитесь в следующих темах.

В ходе выполнения работы были закреплены навыки работы с pandas. Полученная теория была применена на практике. 
Так же была обнаружена особенность в данных, связанная с уникальностью прослушиваемых композиций пользователей. 
Скорее всего это связанно с особеностями подготовки данного DataFrame
**Учебная цель исследования была достигнута**

Проверены три гипотезы. Полученный пайплайн может быть использован при анализе подобных данных для других городов и дней недели. Но перед этим его нужно доработать в соответсвующих местах. **Парктическая цель исследования была достигнута**



<div style="border:solid green 2px; padding: 20px">

<b>Итоговый комментарий ревьюера</b>

Павел, спасибо за проект! Заметно, что ты добросовестно поработал над проектом. Ты хорошо освоил базовый синтаксис Python, умеешь применять базовые методы библиотеки pandas, аккуратно и правильно пишешь код, стараешься писать оптимально. Проект производит хорошее общее впечатление. Ты однозначно движешься в верном направлении!
    
Желтые комментарии — на твое усмотрение: можно поправить сейчас, если есть время/силы/желание, а можно учесть уже в будущих проектах.
    
Я готов принять проект, но отправлю на доработку на случай, если у тебя есть вопросы. Если нет, то приму проект при следующей отправке. 🙂
</div>

<div class="alert alert-info"> <b>Комментарий студента:</b>
    
Саша, спасибо за обратную связь. Приятно с тобой работать. Если будут дополнительные вопросы, замечания, то я с удовльствием их учту. </div>

<div style="border:solid green 2px; padding: 20px">

<b>Итоговый комментарий ревьюера v2</b>

Ты совершенно верно учел все мои замечания. Я дал тебе еще парочку новых комментариев. Надеюсь, все понятно.
    
Я принимаю твой проект и желаю успехов в будущих проектах 🖖
</div>